modules/powers/Hecate.class.php

```diff
@@ -13,10 +13,250 @@ public function __construct($game, $playerId)
+    ];
+    
+    
+    $this->game->notifyPlayer($this->playerId, 'workerMoved', $msg, $args);
+    
+    $args = [
+      'i18n' => ['player_name'],
+      'player_name' => $this->game->getActivePlayerName(),
+      'coords' => $this->game->specialNames['secret']
+    ];
+    
+    $this->game->notifyAllPlayers('message', '${player_name} moves to (${coords})', $args);
+  
+    
+    return true; // do not move again
+  }
+  
+  
+// return null if the logged action is legal wrt the secret workers. Otherwise, return the problematic space
+  public function getAffectedSpaceFromLog($log, $myWorkers)
+  {    
+      if (count($myWorkers) == 0)
+        return null;
+      
+      Utils::filterWorkersById($myWorkers, $log['piece_id'], false);
+  
+      if ($log['action'] == 'move' || $log['action'] == 'force' || $log['action'] == 'build'
+       || $log['action'] == 'placeWorker' || $log['action'] == 'placeToken'
+       || $log['action'] == 'moveToken')
+      { 
+        $args = json_decode($log['action_arg'], true);
+        $space = $args['to'];
+      }
+      elseif ($log['action'] == 'removal')
+      {
+        $space = $this->game->board->getPiece($log['piece_id']);
+      }
+      else
+        return null;
+       
+      if (max(array_map(function($s) use ($space) {
+                      return ($space['x'] == $s['x'] && $space['y'] == $s['y']) ;
+                      } , $myWorkers )))
+         return $space;
+      else
+        return null;
+  
+  }
+  
+  
+  // check if the turn was legal based on Hecate power, and cancel the last actions if necessary
+  // parameter: for Maenads
+  public function endOpponentTurn($testOnly = false) 
+  {
+    $myWorkers = $this->getPlacedWorkers();
+    
+    $logs = $this->game->log->logsForCancelTurn();
+    
+    $space = null;
+    
+    foreach (array_reverse($logs) as $log) {
+        $space = $this->getAffectedSpaceFromLog($log, $myWorkers);
+        if ($space == null)
+          continue;
+        // cancel end of turn
+        
+        if ($testOnly)
+          return true;
+        $moveIds = $this->game->log->cancelTurn($log['log_id']);    
+        break;
+    }
+    
+    if ($testOnly)
+      return false;
+    
+    if ($space == null)
+    {
+      // treat Medusa: kill secret workers only after we know the turn is legal
+      $powers = $this->game->playerManager->getPlayer($this->game->getActivePlayerId())->getPowers();
+      foreach ($powers as $power)
+      {
+        if ($power->getId() != MEDUSA)
+          continue;
+        $argKill = ['workers' => []];
+        $power->argPlayerBuild($argKill, true); // get killable secret workers
+        $power->endPlayerTurn($argKill); // kill them
+      }
+      return;
+    }
+    
+    
+    
+    // display current board state
+    $playerIds = $this->game->playerManager->getPlayerIds();
+    foreach ($playerIds as $playerId) {
+      $this->game->notifyPlayer($playerId, 'cancel', '' , [
```

Comment:
FYI, this will break spectators. Since we do not know the playerId for spectators, the **only** way we communicate with them is via notifyAllPlayers. Probably this means we need to send two notifications -- one public notification to all players, and another private notification to each player with secret pieces.